File: csharp/Primrose.NETStandard2/Grammars/Token.cs
namespace Primrose
{
    /// <summary>
    /// A chunk of text that represents a single element of code,
    /// with fields linking it back to its source.
    /// </summary>
    public class Token
    {
        public string type { get; set; }
        public string value { get; set; }
        public int startStringIndex { get; }

        public int length
        {
            get => value.Length;
        }

        public int endStringIndex
        {
            get => startStringIndex + length;
        }

        public Token(string value, string type, int startStringIndex)
        {
            this.value = value;
            this.type = type;
            this.startStringIndex = startStringIndex;
        }

        public Token clone()
        {
            return new Token(value, type, startStringIndex);
        }

        public Token splitAt(int i)
        {
            var next = value.Substring(i);
            value = value.Substring(0, i);
            return new Token(next, type, startStringIndex + i);
        }

        public override string ToString()
        {
            return $"[{type}: {value}]";
        }
    }
}

File: Primrose.TypeScript/src/Grammars/Token.ts
export type TokenType =
    | "regular"
    | "stringDelim"
    | "newlines"
    | "strings"
    | "startBlockComments"
    | "endBlockComments"
    | "startLineComments"
    | "comments"
    | "lineNumbers"
    | "whitespace"
    | "numbers"
    | "keywords"
    | "operators"
    | "members"
    | "regexes"
    | "functions";

// A chunk of text that represents a single element of code,
// with fields linking it back to its source.
export class Token {
    constructor(
        public value: string,
        public type: TokenType,
        public readonly startStringIndex: number) {
        Object.seal(this);
    }

    get length() {
        return this.value.length;
    }

    get endStringIndex() {
        return this.startStringIndex + this.length;
    }

    clone() {
        return new Token(this.value, this.type, this.startStringIndex);
    }

    splitAt(i: number): Token {
        var next = this.value.substring(i);
        this.value = this.value.substring(0, i);
        return new Token(next, this.type, this.startStringIndex + i);
    }

    toString() {
        return `[${this.type}: ${this.value}]`;
    }
};
